### ISSUE TITLE
Duck array compatibility meeting

### ISSUE THREAD

**Issue by TomNicholas** [pydata/xarray#5648]
## **Proposal: hold a high-level inter-library meeting to sort out roadblocks in the duck-array wrapping efforts.**

Whilst trying to get dask, pint and xarray all working nicely together, I couldn't help but notice there are [important issues](https://github.com/dask/dask/issues/6385) which conclude with a shared sentiment that "we just need to make a decision as to what wraps what" but since then have had essentially no codified consensus, and hence no progress for the past year. Multiply-nested duck-array wrapping is complicated and involves a lot of separate libraries (as this [graph of potential wrappings](https://pint.readthedocs.io/en/stable/numpy.html#Technical-Commentary) shows), but could be an amazingly powerful feature!

![image](https://user-images.githubusercontent.com/35968931/127528046-2999051e-b449-4d0b-a6c7-eaadb091d283.png)

I suggest that as asynchronous discussion hasn't moved this forward, we should instead hold a (hopefully one-off) meeting to make these high-level design decisions.

I'm happy to arrange the meeting, but for this to work we ideally need attendees who understand the issues from the perspective of each of the main libraries involved - some suggestions:
- xarray (@shoyer and @keewis)
- dask (@mrocklin?)
- pint (@jthielen)
- cupy? (@jacobtomlinson?)
- sparse? (@crusaderky?)
- pytorch?? (@rgommers??)

### Possible Agenda (please suggest additions!):

- Which libraries should wrap which other libraries
- Repo/NEP/etc. for standardizing wrapping order and other future decisions
- Outstanding issues to tackle first

### Background reading

- Basic idea of the numpy dispatch mechanism explained in a [blog post](https://blog.christianperone.com/2019/07/numpy-dispatcher-when-numpy-becomes-a-protocol-for-an-ecosystem/)
- @jthielen 's excellent [overview comment](https://github.com/numpy/numpy/pull/16022#issuecomment-655598493), with links to relevant NEP's 
- Pint's [technical commentary](https://pint.readthedocs.io/en/stable/numpy.html#Technical-Commentary) on array type support 

### Some related issues (there are many more - please add)

- https://github.com/pydata/xarray/issues/5559
- https://github.com/pydata/xarray/issues/3950

**Comment by keewis** [comment]
cc @hameerabbasi for `sparse`, and @amcnicho might also be interested

**Comment by mrocklin** [comment]
I would be happy to attend and look forward to what I'm sure will be a vigorous discussion :)  Thank you for providing convenient links to reading materials ahead of time.

As a warning, my responsiveness to github comments these days is not what it used to be.  If I miss something here then please forgive me.

**Comment by hameerabbasi** [comment]
I'd also be happy to attend. Keep in mind I'm in the CET timezone.

**Comment by rgommers** [comment]
I'm happy to join, seems interesting. And yes, I can say something about PyTorch. There probably isn't much to say though - PyTorch is unlikely to adopt `__array_function__` at this point, just like JAX. And it doesn't seem critical for this hierarchy anyway - the fundamental array objects (PyTorch/CuPy/NumPy/Sparse/JAX arrays or tensors) do not have or need a class hierarchy, they are all at the bottom and should not be mixed. 

The key thing here seems to be Dask <-> Xarray <-> Pint, unless I'm missing something?

**Comment by jacobtomlinson** [comment]
Happy to attend. It might also be useful to have @pentschev involved too.

**Comment by TomNicholas** [comment]
Thanks to everyone ho has expressed interest already! I'll give some more time for responses before starting to think about potential meeting times.

> There probably isn't much to say though - PyTorch is unlikely to adopt __array_function__ at this point, just like JAX.

Interesting - could you say a bit more? Looking at these two issues, it seemed more like the question was simply on hold until someone who wanted it badly enough came along?

https://github.com/google/jax/issues/1565

https://github.com/pytorch/pytorch/issues/22402

(I'm interested because I'm in the set of people who have an ambition to use these libraries (wrapped in xarray) at some point further down the line, but that might remain just an ambition :sweat_smile:)

> The key thing here seems to be Dask <-> Xarray <-> Pint, unless I'm missing something?

You're right, that was the primary motivation for this issue, but it also seemed like a good opportunity to get as many libraries talking (and their relationships codified) as possible, especially if one of the outcomes ends up being a type casting hierarchy resource/NEP.

**Comment by jthielen** [comment]
Count me in for the meeting!

---

Here are a few suggestions about possible topics to add to the agenda (based on linked issues/discussions), if we can fit it all in:

- Canonical/minimal API of a "duck array" and how to detect it (though may be superseded by NEPs [30](https://numpy.org/neps/nep-0030-duck-array-protocol.html) and [47](https://numpy.org/neps/nep-0047-array-api-standard.html) among others)
- Consistency of type handling behavior (i.e., between construction, binary ops, `__array_ufunc__`, `__array_function__`, and array modules...for example, these are uniform in Pint, but construction and array module functions are deliberately different from the others for Dask arrays)
- API for inter-type casting and changing what types are used in a nested array (e.g. #3245 and #5568)
- How to handle unknown duck arrays
- Nested array reprs (both short and full)

---

Also, tagging a few other array type libraries and maintainers/contributors who may be interested (please ping the relevant folks if you know them):

- unyt (@ngoldbaum)
- astropy.units (??)
- NumPy masked arrays (??)

(interesting side note is that these are all ndarray subclasses right now...perhaps discussing the interplay between array subclassing and wrapping is in order too?)

**Comment by benbovy** [comment]
- Awkward? (@jpivarski)

**Comment by TomNicholas** [comment]
> NumPy masked arrays (??)

Tagging @greglucas because of his work on https://github.com/numpy/numpy/pull/16022#issuecomment-888495182 - at least I think a very common desired use case of multi-nested arrays would be `xarray(pint(dask(np.masked))`.

**Comment by pentschev** [comment]
I'm also happy to join the meeting. Thanks @TomNicholas for the initiative here and @jacobtomlinson for tagging me.

**Comment by greglucas** [comment]
Happy to hop on a call for this as well, thanks for organizing all of this @TomNicholas!

**Comment by jpivarski** [comment]
> * Awkward? (@jpivarski)

I'm interested. Let us know when the time will be or if there's a poll for picking a time. Thanks!

**Comment by rgommers** [comment]
> Interesting - could you say a bit more? Looking at these two issues, it seemed more like the question was simply on hold until someone who wanted it badly enough came along?

There is a significant backwards compatibility break when a library adds `__array_ufunc__` and `__array_function__`. JAX maintainers were not comfortable with that. @shoyer wrote https://numpy.org/neps/nep-0037-array-module.html as a follow-up largely because of that. That NEP is effectively superceded by the array API standard (https://data-apis.org/array-api/latest/ and NEP 47). PyTorch has decided to adopt that and implementation is well underway. Experimental support in NumPy will land next week (complete except for `linalg`). CuPy and MXNet plan to follow that NumPy implementation. JAX and Dask not yet decided I believe, but likely to follow NumPy as well.

> Canonical/minimal API of a "duck array" and how to detect it (though may be superseded by NEPs 30 and 47 among others)

This is basically what the array API standard provides (most functions follow NumPy, with deviations mostly where other libraries were already deviating because they could implement something on GPU, with a JIT, or with a non-strided memory model). `__array_function__` has worked quite well for CuPy and Dask, because they follow the NumPy API almost to the letter, with only a couple of exceptions (e.g. 0-D array instead of array scalars in CuPy). JAX, PyTorch and MXNet all deviate much more, and since the NumPy API is not very well-defined (there's 1500+ public objects plus more semi-public ones), you'd have no guarantees about what works and what doesn't.

That said, `__array_ufunc__` and `__array_function__` aren't going anywhere. The RAPIDS ecosystem is invested in it and I believe largely happy with it. So adding Xarray and Pint to the mix sounds potentially interesting.

**Comment by SimonHeybrock** [comment]
>     * scipp (@SimonHeybrock, xref [NEP 18, physical units, uncertainties, and the scipp library? #3509](https://github.com/pydata/xarray/issues/3509))

Thanks! I am definitely interested.

**Comment by khaeru** [comment]
👂🏾

**Comment by TomNicholas** [comment]
Apologies for the long delay, but I would like to try to organise this meeting for some time in the next few weeks. If you are interested could you please fill out this [When2Meet meeting time poll](https://www.when2meet.com/?12888573-1TQNs). (If none of these times work for you because of time zone issues or something then please say so!) It would be great to have lots of libraries represented.

[Agenda doc here](https://docs.google.com/document/d/1SIHECqe18Cp9WOe8dbGH8bnpPeraOd4IvNB1cRsntys/edit?usp=sharing) - I will come back to this in more detail soon, but if you have specific things you would like to discuss then please add them.

**Comment by TomNicholas** [comment]
TOMORROW: So there is just one slot when essentially everyone said they were free - but it's 11:00am EDT September 22nd, i.e. tomorrow morning.

I appreciate that's late notice - but could we can try for that and if we only get the super keen beans attending this time then we woud still be able to have a useful initial discussion about the exact problems that need resolving?

Alternatively if people could react to this comment with thumbs up / down for "that time is still good" / "no I need more notice" then that would be useful.

@mrocklin are you or anyone else who is able to speak for dask free? I noticed that @jacobtomlinson put down that you were free tomorrow also?

**Comment by mrocklin** [comment]
Surprisingly I happen to be free tomorrow at exactly that time.  I've
blocked it off.  If you want to send a calendar invite to mrocklin at
coiled that would be welcome.

**Comment by jacobtomlinson** [comment]
I can be there!

**Comment by jakirkham** [comment]
Maybe too soon to ask, but do we have a link to the video call?

**Comment by TomNicholas** [comment]
> Surprisingly I happen to be free tomorrow at exactly that time.

Great! I sent an invite to the few people whose emails I could easily find - let me know if that didn't arrive @mrocklin .

> Maybe too soon to ask, but do we have a link to the video call?

I think we can probably just use the one we use for the [xarray bi-weekly dev meetings](https://github.com/pydata/xarray/issues/4001):

https://us02web.zoom.us/j/88251613296?pwd=azZsSkU1UWJZTVFKNnhIUVdZcENUZz09

**Comment by crusaderky** [comment]
I'd like to attend too

**Comment by beckernick** [comment]
I'd also like to attend, primarily just to learn.

**Comment by leofang** [comment]
Will try to join.

**Comment by hodgestar** [comment]
I'd like to attend on behalf of QuTiP. I'll likely mostly listen -- QuTiP is not directly affected in the way that xarray, Dask, cupy, etc are, but we're users of potentially all of these array types (and already explicitly support numpy, CuPy, and our own sparse array format) and we are facing similar issues of our own (i.e. users of QuTiP are asking us to develop a `__qutip_qobj__` style protocol similar to numpy's and we would like to learn the lessons of the last decade of numpy rather than repeat the steps ourselves in the coming one).

**Comment by TomNicholas** [comment]
Change of plan - use my personal zoom room instead (as I don't think I have permissions to start the xarray one):

https://columbiauniversity.zoom.us/j/97399260034?pwd=WS9GUnVrcG1YSWRPeXhITXRhdEJZUT09

Also unless anyone has a problem I'm going to record the meeting, just for note-taking purposes.

**Comment by hameerabbasi** [comment]
I would very much prefer not to be recorded.

**Comment by mrocklin** [comment]
It looks like there are some other Dask folks participating.  I'll step
back and let them take over on our end.

**Comment by rgommers** [comment]
There are also some very interesting PyTorch development discussion, there's a lot of Tensor subclasses in the making and thought being put into how those interact:
- https://dev-discuss.pytorch.org/t/state-of-pytorch-core-september-2021-edition/332#alternative-tensors-5
- https://pytorch-dev-podcast.simplecast.com/episodes/tensor-subclasses-and-liskov-substitution-principle

**Comment by TomNicholas** [comment]
Thank you very much to everyone who came today (all 18 of us!). 

I've raised the ToDos we had as an [issue](https://github.com/pydata/duck-array-discussion/issues/1) in the [new duck array discussion repo](https://github.com/pydata/duck-array-discussion) - anyone interested in this topic might want to watch the activity on that repository.

The plan now is to continue via asynchronous discussion on that new repo, referring back to issues in specific libraries when necessary. If we come to another sticking point that requires in-person discussion then we can organise another face-to-face meeting at that point, which I am happy to do. However we won't organise another meeting until there is a specific blockage to discuss. I will therefore keep this issue open so I can use it to alert anyone who might be interested in a follow-up meeting.
